### PR TITLE
[JUJU-4151] - Fix/lp 2025960

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR adds secret support for the k8s controller model.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

```sh
juju switch k1:controller
k1 (controller) -> k1:admin/controller

juju exec --unit controller/0 -- secret-add foo=bar
secret://a38b6cd2-b37d-4766-890d-acc62e414139/cijsiv6ffbas7adlb97g

juju exec --unit controller/0 -- secret-add foo=bar
secret://a38b6cd2-b37d-4766-890d-acc62e414139/cijsj0mffbas7adlb98g

juju exec --unit controller/0 -- secret-get cijsj0mffbas7adlb98g
foo: bar

juju exec --unit controller/0 -- secret-get cijsiv6ffbas7adlb97g
foo: bar

juju add-model t1
Added 't1' model on microk8s/localhost with credential 'microk8s' for user 'admin'

juju deploy snappass-test
Located charm "snappass-test" in charm-hub, revision 9
Deploying "snappass-test" from charm-hub charm "snappass-test", revision 9 in channel stable on ubuntu@20.04/stable

juju exec --unit snappass-test/0 -- secret-add foo=bar
secret://8d6b9148-8ed7-4461-89d8-c549fbdadc7f/cijsjbuffbas7adlb990

juju exec --unit snappass-test/0 -- secret-add foo=bar
secret://8d6b9148-8ed7-4461-89d8-c549fbdadc7f/cijsjcmffbas7adlb99g

juju exec --unit snappass-test/0 -- secret-get cijsjbuffbas7adlb990
foo: bar

juju exec --unit snappass-test/0 -- secret-get cijsjcmffbas7adlb99g
foo: bar

```

## Documentation changes

No

## Bug reference

https://bugs.launchpad.net/juju/+bug/2025960

### CODE REVIEW

**Comment by hpidcock** [comment]
```
----------------------------------------------------------------------
FAIL: rbac_test.go:264: rbacSuite.TestEnsureSecretAccessTokeControllerModelUpdate

[LOG] 0:00.002 DEBUG juju.kubernetes.provider updating service account "unit-gitlab-0"
rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:183:
    h.t.Fatalf(format, args...)
... Error: Unexpected call to *mocks.MockClusterRoleInterface.Get([context.TODO controller-k1-unit-gitlab-0 {{ } }]) at /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac.go:438 because: 
expected call at /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:326 doesn't match the argument at index 1.
Got: controller-k1-unit-gitlab-0 (string)
Want: is equal to unit-gitlab-0 (string)

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:180:
    h.t.Errorf(format, args...)
... Error: missing call(s) to *mocks.MockClusterRoleInterface.Get(is anything, is equal to unit-gitlab-0 (string), is equal to {{ } } (v1.GetOptions)) /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:326

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:180:
    h.t.Errorf(format, args...)
... Error: missing call(s) to *mocks.MockClusterRoleBindingInterface.Get(is anything, is equal to controller-k1-unit-gitlab-0 (string), is equal to {{ } } (v1.GetOptions)) /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:328

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:180:
    h.t.Errorf(format, args...)
... Error: missing call(s) to *mocks.MockServiceAccountInterface.CreateToken(is anything, is equal to unit-gitlab-0 (string), is equal to &TokenRequest{ObjectMeta:{      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:TokenRequestSpec{Audiences:[],BoundObjectRef:nil,ExpirationSeconds:*600,},Status:TokenRequestStatus{Token:,ExpirationTimestamp:0001-01-01 00:00:00 +0000 UTC,},} (*v1.TokenRequest), is equal to {{ } []  } (v1.CreateOptions)) /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:332

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:180:
    h.t.Errorf(format, args...)
... Error: missing call(s) to *mocks.MockClusterRoleBindingInterface.Patch(is anything, is equal to unit-gitlab-0 (string), is equal to application/strategic-merge-patch+json (types.PatchType), is anything, is equal to {{ } [] <nil> juju } (v1.PatchOptions)) /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:329

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:180:
    h.t.Errorf(format, args...)
... Error: missing call(s) to *mocks.MockClusterRoleInterface.Update(is anything, is equal to &ClusterRole{ObjectMeta:{controller-k1-unit-gitlab-0  controller-k1    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[app.kubernetes.io/managed-by:juju app.kubernetes.io/name:gitlab] map[] [] []  []},Rules:[]PolicyRule{PolicyRule{Verbs:[create patch],APIGroups:[*],Resources:[secrets],ResourceNames:[],NonResourceURLs:[],},PolicyRule{Verbs:[get list],APIGroups:[*],Resources:[namespaces],ResourceNames:[],NonResourceURLs:[],},},AggregationRule:nil,} (*v1.ClusterRole), is equal to {{ } []  } (v1.UpdateOptions)) /home/jenkins/go/src/github.com/juju/juju/caas/kubernetes/provider/rbac_test.go:327

rbac_test.go:337:
    token, err := s.broker.EnsureSecretAccessToken(tag, nil, nil, nil)
/home/jenkins/go/pkg/mod/github.com/golang/mock@v1.6.0/gomock/controller.go:183:
    h.t.Fatalf(format, args...)
... Error: aborting test due to missing call(s)


----------------------------------------------------------------------
```

File: caas/kubernetes/provider/rbac.go [hunk 1]
```diff
@@ -659,21 +673,90 @@ func (k *kubernetesClient) EnsureSecretAccessToken(tag names.Tag, owned, read, r
 	return tr.Status.Token, nil
 }
 
-func (k *kubernetesClient) ensureRoleForSecretAccessToken(objMeta v1.ObjectMeta, owned, read, removed []string) (*rbacv1.Role, error) {
+func (k *kubernetesClient) ensureBindingForSecretAccessToken(sa *core.ServiceAccount, objMeta v1.ObjectMeta, owned, read, removed []string) error {
+	ctx := context.TODO()
+	isControllerModel := k.Config().Name() == environsbootstrap.ControllerModelName
+
+	if isControllerModel {
```
Comment (wallyworld) [inline caas/kubernetes/provider/rbac.go:680]: please move this block to a separate function

File: caas/kubernetes/provider/rbac.go [hunk 2]
```diff
@@ -710,6 +790,22 @@ func rulesForSecretAccess(namespace string, existing []rbacv1.PolicyRule, owned,
...
 		}
+		if isControllerModel {
+			// We need to be able to list/get all namespaces for units' in controller model.
+			existing = append(existing, rbacv1.PolicyRule{
+				APIGroups: []string{rbacv1.APIGroupAll},
+				Resources: []string{"namespaces"},
+				Verbs:     []string{"get", "list"},
+			})
+		} else {
+			// We just need to be able to list/get our own namespace for units' in other models.
```
Comment (wallyworld) [inline caas/kubernetes/provider/rbac.go:801]: remove ' (and above)

**Review by wallyworld — APPROVED** [review]
Thank you. Just needs a little clean up.

**Pull request merged**